dex/ws/wslink_test.go

```diff
@@ -0,0 +1,199 @@
+}
+
+func (c *ConnStub) Close() error {
+	// make ReadMessage return with a close error
+	atomic.StoreInt32(&c.closed, 1)
+	c.inErr <- &websocket.CloseError{
+		Code: websocket.CloseNormalClosure,
+		Text: "bye",
+	}
+	return nil
+}
+func (c *ConnStub) SetReadDeadline(t time.Time) error {
+	return nil
+}
+func (c *ConnStub) SetWriteDeadline(t time.Time) error {
+	return nil
+}
+func (c *ConnStub) ReadMessage() (int, []byte, error) {
+	if atomic.LoadInt32(&c.closed) == 1 {
+		return 0, nil, &websocket.CloseError{
+			Code: websocket.CloseAbnormalClosure,
+			Text: io.ErrUnexpectedEOF.Error(),
+		}
+	}
+	select {
+	case msg := <-c.inMsg:
+		return len(msg), msg, nil
+	case err := <-c.inErr:
+		return 0, nil, err
+	}
+}
+
+const (
+	stdDev = 500.0
+	min    = int64(50)
+)
+
+func microSecDelay(stdDev float64, min int64) time.Duration {
+	return time.Microsecond * time.Duration(int64(stdDev*math.Abs(rand.NormFloat64()))+min)
+}
+
+var lastID int64 = -1 // first msg.ID should be 0
+
+func (c *ConnStub) WriteMessage(_ int, b []byte) error {
+	if atomic.LoadInt32(&c.closed) == 1 {
+		return websocket.ErrCloseSent
+	}
+	msg, err := msgjson.DecodeMessage(b)
+	if err != nil {
+		return err
+	}
+	if msg.ID != uint64(lastID+1) {
+		return fmt.Errorf("sent out of sequence. got %d, want %v", msg.ID, uint64(lastID+1))
+	}
+	lastID++
+	//writeDuration := time.Microsecond * time.Duration(1+rand.Int63n(40000))
+	writeDuration := microSecDelay(stdDev, min)
+	//fmt.Printf(" <ConnStub> writing message in %v: %v\n", writeDuration, string(b))
+	time.Sleep(writeDuration)
+	return nil
+}
+func (c *ConnStub) WriteControl(messageType int, data []byte, deadline time.Time) error {
+	switch messageType {
+	case websocket.PingMessage:
+		fmt.Println(" <ConnStub> ping sent to peer")
+	case websocket.TextMessage:
+		fmt.Println(" <ConnStub> message sent to peer:", string(data))
+	case websocket.CloseMessage:
+		fmt.Println(" <ConnStub> close control frame sent to peer")
+	default:
+		fmt.Printf(" <ConnStub> message type %d sent to peer\n", messageType)
+	}
+	return nil
+}
+
+func TestWSLink_send(t *testing.T) {
+	//rand.Seed(int64(time.Now().Nanosecond()))
+	backendLogger := slog.NewBackend(os.Stdout)
+	defer os.Stdout.Sync()
+	log := backendLogger.Logger("ws_TEST")
+	log.SetLevel(slog.LevelTrace)
+	UseLogger(log)
+
+	inMsgHandler := func(msg *msgjson.Message) *msgjson.Error {
+		return nil
+	}
+
+	conn := &ConnStub{
+		inMsg: make(chan []byte, 1),
+		inErr: make(chan error, 1),
+	}
+	wsLink := NewWSLink("127.0.0.1", conn, time.Second, inMsgHandler)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// start the in/out/pingHandlers, and the initial read deadline
+	wg, err := wsLink.Connect(ctx)
+	if err != nil {
+		t.Fatalf("Connect: %v", err)
+	}
+
+	defer wg.Wait()
+
+	// hit the inHandler once before testing sends
+	msg, _ := msgjson.NewRequest(12, "beep", "boop")
+	b, err := json.Marshal(msg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	conn.inMsg <- b
```

Comment:
Nothing actually be tested here?

Reply:
Kind of a silly one, but this was really just testing that `(*WSLink).inHandler` works OK and calls the handler.  I forgot to update `inMsgHandler` to do something to confirm that though... fixing.